File: packages/go/validation.go
package connparse

import (
	"regexp"
	"strings"
)

type validationResult struct {
	errors   []Diagnostic
	warnings []Diagnostic
}

func validateAddress(address *Address, definition Definition, options Options) validationResult {
	errors := []Diagnostic{}
	warnings := []Diagnostic{}
	validation := definition.Validation

	hasHost := asString(address.Authority["host"]) != ""
	hasHosts := false
	if hosts, ok := address.Authority["hosts"].([]map[string]any); ok {
		hasHosts = len(hosts) > 0
	}

	if validation.RequireHost && !hasHost && !hasHosts {
		errors = append(errors, diagnostic("MISSING_HOST", displayName(definition)+" requires a host", "authority"))
	}
	if definition.ResourceRule.Required && address.Resource.Name == nil {
		errors = append(errors, diagnostic("MISSING_RESOURCE", displayName(definition)+" requires a resource", "resource.name"))
	}
	if definition.PathRule.Required && address.Path == "" {
		errors = append(errors, diagnostic("MISSING_PATH", displayName(definition)+" requires a path", "path"))
	}

	rangeRule := validation.PortRange
	if rangeRule.Min != 0 || rangeRule.Max != 0 {
		for _, port := range collectPorts(address.Authority) {
			if port < rangeRule.Min || port > rangeRule.Max {
				errors = append(errors, diagnostic("INVALID_PORT", "Port must be between "+toString(rangeRule.Min)+" and "+toString(rangeRule.Max), "authority.port"))
			}
		}
	}

	for key, value := range address.Query {
		rule, ok := definition.QueryParameters[key]
		if !ok {
			item := diagnostic("UNKNOWN_QUERY_PARAMETER", key+" is not declared for "+definition.ID, "query."+key)
			if options.Strict {
				errors = append(errors, item)
			} else {
				warnings = append(warnings, item)
			}
			continue
		}
		errors = append(errors, validateQueryValue(rule, key, value)...)
	}

	return validationResult{errors: errors, warnings: warnings}
}

func displayName(def Definition) string {
	if def.Name != "" {
		return def.Name
	}
	return def.ID
}

func collectPorts(authority map[string]any) []int {
	ports := []int{}
	if port, ok := toInt(authority["port"]); ok {
		ports = append(ports, port)
	}
	if hosts, ok := authority["hosts"].([]map[string]any); ok {
		for _, host := range hosts {
			if port, ok := toInt(host["port"]); ok {
				ports = append(ports, port)
			}
		}
	}
	return ports
}

func validateQueryValue(rule QueryRule, key string, value any) []Diagnostic {
	errors := []Diagnostic{}
	for _, item := range valuesFor(value) {
		if rule.Type == "boolean" && !isBooleanString(item) {
			errors = append(errors, diagnostic("INVALID_QUERY_PARAMETER_TYPE", key+" must be a boolean", "query."+key))
		}
		if rule.Type == "number" && !isNumberString(item) {
			errors = append(errors, diagnostic("INVALID_QUERY_PARAMETER_TYPE", key+" must be a number", "query."+key))
		}
		if len(rule.Allowed) > 0 && !allowedContains(rule.Allowed, item) {
			errors = append(errors, diagnostic("INVALID_QUERY_PARAMETER_VALUE", key+" must be one of: "+joinAllowed(rule.Allowed), "query."+key))
		}
	}
	return errors
}

func valuesFor(value any) []any {
	if values, ok := value.([]any); ok {
		return values
	}
	return []any{value}
}

func isBooleanString(value any) bool {
	switch strings.ToLower(toString(value)) {
	case "true", "false", "1", "0", "yes", "no":
		return true
	default:
		return false
	}
}

func isNumberString(value any) bool {
	return regexp.MustCompile(`^-?\d+(\.\d+)?$`).MatchString(toString(value))
}

func allowedContains(allowed []any, value any) bool {
	text := toString(value)
	for _, item := range allowed {
		if toString(item) == text {
			return true
		}
	}
	return false
}

func joinAllowed(allowed []any) string {
	parts := make([]string, len(allowed))
	for i, item := range allowed {
		parts[i] = toString(item)
	}
	return strings.Join(parts, ", ")
}

func toInt(value any) (int, bool) {
	switch v := value.(type) {
	case int:
		return v, true
	case int64:
		return int(v), true
	case float64:
		return int(v), v == float64(int(v))
	default:
		return 0, false
	}
}

File: packages/js/src/validation.js
import { diagnostic } from './diagnostics.js';

function valuesFor(value) {
  return Array.isArray(value) ? value : [value];
}

function normalizeLookupKey(value) {
  return String(value ?? '').trim().toLowerCase();
}

function isBooleanString(value) {
  return ['true', 'false', '1', '0', 'yes', 'no'].includes(String(value).toLowerCase());
}

function isNumberString(value) {
  return /^-?\d+(\.\d+)?$/.test(String(value));
}

function normalizeBoolean(value) {
  const text = String(value).toLowerCase();
  if (['true', '1', 'yes'].includes(text)) return 'true';
  if (['false', '0', 'no'].includes(text)) return 'false';
  return String(value);
}

function normalizeNumber(value) {
  const number = Number(value);
  return Number.isFinite(number) ? String(number) : String(value);
}

function mappedValue(value, map) {
  if (map == null || typeof map !== 'object') {
    return undefined;
  }

  const entries = new Map(
    Object.entries(map).map(([key, item]) => [normalizeLookupKey(key), item])
  );
  return entries.get(normalizeLookupKey(value));
}

function queryRuleForKey(key, definition) {
  const normalizedKey = normalizeLookupKey(key);

  for (const [canonicalKey, rule] of Object.entries(definition?.query_parameters || {})) {
    if (normalizeLookupKey(canonicalKey) === normalizedKey) {
      return { canonicalKey, rule };
    }

    for (const alias of rule?.aliases || []) {
      if (normalizeLookupKey(alias) === normalizedKey) {
        return { canonicalKey, rule };
      }
    }
  }

  return {
    canonicalKey: String(key),
    rule: null
  };
}

function normalizeValueForRule(rule, value) {
  let normalizedValue = String(value);
  if (rule.type === 'boolean') {
    normalizedValue = normalizeBoolean(value);
  } else if (rule.type === 'number') {
    normalizedValue = normalizeNumber(value);
  }

  const mapped = mappedValue(normalizedValue, rule.normalized_values);
  return mapped == null ? normalizedValue : String(mapped);
}

function validateQueryValue(rule, key, value) {
  const errors = [];
  for (const item of valuesFor(value)) {
    if (rule.type === 'boolean' && !isBooleanString(item)) {
      errors.push(diagnostic('INVALID_QUERY_PARAMETER_TYPE', `${key} must be a boolean`, `query.${key}`));
    }
    if (rule.type === 'number' && !isNumberString(item)) {
      errors.push(diagnostic('INVALID_QUERY_PARAMETER_TYPE', `${key} must be a number`, `query.${key}`));
    }

    const normalizedItem = normalizeValueForRule(rule, item);
    if (
      Array.isArray(rule.allowed) &&
      !rule.allowed.some((allowed) => normalizeValueForRule(rule, allowed) === normalizedItem)
    ) {
      errors.push(
        diagnostic(
          'INVALID_QUERY_PARAMETER_VALUE',
          `${key} must be one of: ${rule.allowed.join(', ')}`,
          `query.${key}`
        )
      );
    }
  }
  return errors;
}

export function validateAddress(address, definition, options = {}) {
  const errors = [];
  const warnings = [];
  const validation = definition.validation || {};

  const hasHost = Boolean(address.authority.host);
  const hasHosts = Array.isArray(address.authority.hosts) && address.authority.hosts.length > 0;

  if (validation.require_host && !hasHost && !hasHosts) {
    errors.push(diagnostic('MISSING_HOST', `${definition.name || definition.id} requires a host`, 'authority'));
  }

  if (definition.resource?.required && !address.resource.name) {
    errors.push(
      diagnostic('MISSING_RESOURCE', `${definition.name || definition.id} requires a resource`, 'resource.name')
    );
  }

  if (definition.path?.required && !address.path) {
    errors.push(diagnostic('MISSING_PATH', `${definition.name || definition.id} requires a path`, 'path'));
  }

  const range = validation.port_range;
  if (range) {
    const ports = [];
    if (address.authority.port != null) ports.push(address.authority.port);
    if (Array.isArray(address.authority.hosts)) {
      for (const host of address.authority.hosts) {
        if (host.port != null) ports.push(host.port);
      }
    }
    for (const port of ports) {
      if (!Number.isInteger(port) || port < range.min || port > range.max) {
        errors.push(diagnostic('INVALID_PORT', `Port must be between ${range.min} and ${range.max}`, 'authority.port'));
      }
    }
  }

  for (const [key, value] of Object.entries(address.query || {})) {
    const { canonicalKey, rule } = queryRuleForKey(key, definition);
    if (!rule) {
      const item = diagnostic('UNKNOWN_QUERY_PARAMETER', `${key} is not declared for ${definition.id}`, `query.${key}`);
      if (options.strict) errors.push(item);
      else warnings.push(item);
      continue;
    }
    errors.push(...validateQueryValue(rule, canonicalKey, value));
  }

  return { errors, warnings };
}
